### ISSUE TITLE
twisted.web.template documentation doesn't cover very basic use-case of rendering more than one of something

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5286]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#5286|
|Type|enhancement|
|Created|2011-09-29 17:32:48Z|
|Branch|https://github.com/twisted/twisted/tree/template-docs-5286-2|

The [twisted.web.template narrative documentation](http://twistedmatrix.com/documents/current/web/howto/twisted-templates.html) covers only the very basic use-case of putting one thing onto a web page by filling a slot inside a renderer.

However, in pretty much every application that wants to produce HTML, there will very quickly be a need to repeat a node.

The way you do this in twisted.web.template is to repeatedly _yield_ the node rather than to simply _return_ it.  Let's say you have a template like:

```
#!text/html
<html ...>
<body>
<table>
<tr t:render="row">
<td><t:slot name="a" /></td>
<td><t:slot name="b" /></td>
</tr>
</table>
</body>
</body>
```
you could fill it out with a renderer like this:
```python
@renderer
def row(self, request, tag):
    for datum in self.data:
        yield tag.clone().fillSlots(a=datum.a, b=datum.b)
```
Note that the relevant renderer is on the _row_, not the table, because it is the row which is repeated.

Attachments:

* [template-docs-5286.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/929/929372690258622726d738fe0c66d2cdf93df14c/c64288a4a8962e7b2bf9cdf6ca0f65cbaed3507d.patch) (3105 bytes) - added by indigo on 2011-09-30 21:33:38Z - documentation addition
* [template-docs-5286.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/929/929372690258622726d738fe0c66d2cdf93df14c/9e9e9293c725f26381f0ea6baef1d5811143bfff.patch) (23460 bytes) - added by indigo on 2011-10-03 13:59:14Z - with news file

<details><summary>Searchable metadata</summary>

```
trac-id__5286 5286
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_template_docs_5286_2 branches/template-docs-5286-2
branch_author__indigo indigo
status__closed closed
resolution__fixed fixed
component__web web
keywords__documentation_easy documentation easy
time__1317317568000000 1317317568000000
changetime__1317767728000000 1317767728000000
version__None None
owner__exarkun exarkun
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)|@bitglue set owner to @bitglue<br>@bitglue set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_3"></a>|@bitglue removed owner<br>@bitglue set status to `new`|
|-|-|

ready for review. I also took this opportunity to remove the "A nice example of Slots and Attributes" section, since it seemed redundant with the "Slots" section two lines above it.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

(In [32717]) Branching to 'template-docs-5286'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/template-docs-5286)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Perhaps more relevant to this particular ticket: [lore2sphinx root](http://buildbot.twistedmatrix.com/builds/sphinx-html/447-16002/)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to @bitglue|
|-|-|

Looks like the famous whitespace bugs in https://bitbucket.org/khorn/lore2sphinx have bit us again.  You'll notice that [the modified sections in the lore2sphinx output](http://buildbot.twistedmatrix.com/builds/sphinx-html/447-16002/projects/web/howto/twisted-templates.html) have strange stuff in them now.

I took the liberty of rearranging the whitespace in r32722 - in case we're stuck with this for a while, one general rule with l2s is that you can't put a `<code>` tag at the beginning of a line, except at the start of a paragraph.  You can see the [updated build results here](http://buildbot.twistedmatrix.com/builds/sphinx-html/449-16003/projects/web/howto/twisted-templates.html).  You might want to check them over to make sure that I didn't edit anything but the whitespace :).

If you can provide a NEWS file, I think this can be ready to land.  (Personally I find that it's helpful to write the NEWS file first, so as to keep the ticket's purpose firmly in mind while working on it; and then it's harder to forget that step.)

**Comment by twisted-trac** [comment]
|[<img alt="bitglue's avatar" src="https://avatars.githubusercontent.com/u/1232472?s=50" width="50" height="50">](https://github.com/bitglue)<a name="note_8"></a>|@bitglue removed owner|
|-|-|

new patch is relative to trunk, adds a news file, and includes changes in svn. No other changes.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [32752]) Branching to 'template-docs-5286-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [32753]) Apply just the parts of template-docs-5286.2.patch that seem appropriate

refs [#5286](https://github.com/twisted/twisted/issues/5286)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun|
|-|-|

Patch has some extra junk in it :/  Talk to me on IRC about getting direct SVN access.  Should make things easier for everyone.

Looking at just the parts that seem related, looks good to me, and the sphinx conversion seems happy now as well.  Thanks, I'll merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [32754]) Merge template-docs-5286-2

Author: indigo
Reviewer: glyph, exarkun
Fixes: [#5286](https://github.com/twisted/twisted/issues/5286)

Add a section to the Twisted Web documentation demonstrating how sequences
of data can be rendered using Twisted Web's templating features.
